_source/_posts/2021-09-30-oauth-sessions-with-java.adoc

```diff
@@ -0,0 +1,316 @@
+---
+layout: blog_post
+title: "Session Clustering for OAuth 2.0 Applications"
+author: brian-demers
+by: advocate
+communities: [java]
+description: "Learn how sessions are used with OAuth 2.0 and build an example with HAproxy, Redis, and Spring Boot."
+tags:  [spring-session, spring, microservices, java, spring-boot, haproxy, oauth2, oidc, session]
+tweets:
+- "Apps using OIDC/OAuth 2.0 don't have 'special' sessions requirements, but you still need to think about them 🧠"
+- "Confused how #OAuth2 and OIDC affect your web app's sessions? Learn more and build an example with #Redis, #HAproxy, and #SpringBoot 💡"
+- "Scaling an app from 1 to N means you need to think about load balancing, session management, and how those relate to Auth🔒"
+image: blog/oauth-sessions-with-java/social.png
+type: conversion
+---
+
+:toc: macro
+:page-liquid:
+:experimental:
+
+A common OAuth 2.0 question we get: "How do I deal with OAuth in a load-balanced application?" The short answer: There's nothing specific about session clustering for OAuth. The longer answer is—you likely still need to worry about cluster session management. This post will discuss how an OAuth login relates to your application's session. And we'll build a simple, secure, load-balanced application to demonstrate.
+
+toc::[]
+
+**Prerequisites**
+
+- https://docs.docker.com/get-docker/[Docker 20.10+]
+- https://adoptopenjdk.net/?variant=openjdk11[Java 11+], use https://sdkman.io/[SDKMAN] to manage multiple Java versions!
+- https://httpie.io/download[HTTPie], a user-friendly HTTP CLI
+- https://cli.okta.com/[Okta CLI]
+
+== Sessions and OAuth 2.0 Applications
+
+An application that uses an OAuth 2.0 redirect, an https://oauth.net/2/grant-types/authorization-code/["Authorization Code Grant"], typically uses a server-side session to temporarily store the state of the login process until it completes. Let's look at a simplified OAuth login flow:
+
+image::{% asset_path 'blog/oauth-sessions-with-java/oauth-code-flow.png' %}[alt=Login Flow Sequence Diagram,width=800,align=center]
+
+1. A user requests to log in - either by clicking on a login button or automatically when they request a protected page.
+2. Web App stores information about the current session - OAuth state info and optionally a PKCE code verifier and/or a nonce for OpenId Connect (OIDC).
+3. Respond with a browser redirect to the authorization server.
```

Comment:
```suggestion
3. The response is a browser redirect to the authorization server.
```
so that each step in the numbered series has the same sentence pattern, so it's easier to follow...